ts/src/bitget.ts

```diff
@@ -3294,51 +3295,96 @@ export default class bitget extends Exchange {
         const marketType = market['spot'] ? 'spot' : 'swap';
         const timeframes = this.options['timeframes'][marketType];
         const selectedTimeframe = this.safeString (timeframes, timeframe, timeframe);
+        const duration = this.parseTimeframe (timeframe) * 1000;
         const request = {
             'symbol': market['id'],
             'granularity': selectedTimeframe,
         };
-        const until = this.safeInteger2 (params, 'until', 'till');
-        params = this.omit (params, [ 'until', 'till' ]);
+        const defaultLimit = 100; // by default, exchange returns 100 items
+        const msInDay = 1000 * 60 * 60 * 24;
         if (limit !== undefined) {
+            limit = Math.min (limit, maxLimit);
             request['limit'] = limit;
         }
-        if (since !== undefined) {
-            request['startTime'] = since;
+        const until = this.safeInteger2 (params, 'until', 'till');
+        params = this.omit (params, [ 'until', 'till' ]);
+        if (until !== undefined) {
+            request['endTime'] = until;
         }
         if (since !== undefined) {
-            if (limit === undefined) {
-                limit = 100; // exchange default
+            request['startTime'] = since;
+            if (market['spot'] && (until === undefined)) {
+                // for spot we need to send "entTime" too
+                const limitForEnd = (limit !== undefined) ? limit : defaultLimit;
+                const calculatedEnd = this.sum (since, duration * limitForEnd);
+                request['endTime'] = calculatedEnd;
             }
-            const duration = this.parseTimeframe (timeframe) * 1000;
-            request['endTime'] = this.sum (since, duration * (limit + 1)) - 1;  // limit + 1)) - 1 is needed for when since is not the exact timestamp of a candle
-        } else if (until !== undefined) {
-            request['endTime'] = until;
-        } else {
-            request['endTime'] = this.milliseconds ();
         }
         let response = undefined;
-        const thirtyOneDaysAgo = this.milliseconds () - 2678400000;
+        const now = this.milliseconds ();
+        // retrievable periods listed here:
+        // - https://www.bitget.com/api-doc/spot/market/Get-Candle-Data#request-parameters
+        // - https://www.bitget.com/api-doc/contract/market/Get-Candle-Data#description
+        const retrievableDaysMapForSpot = {
+            '1m': 30,
+            '3m': 30,
+            '5m': 30,
+            '10m': 52,
+            '15m': 52,
+            '30m': 52,
+            '1h': 83,
+            '2h': 120,
+            '4h': 240,
+            '6h': 360,
+            '12h': 360,
+            '1d': 360,
+            '3d': 1000,
+            '1w': 1000,
+            '1M': 1000,
```

Comment:
1) i don't think its place should be in options, because they are not actually changeable "options" per se for user supposed to change/override, it's hardlimits given by exchange. but ok, i'll move them into options.
2)  this is the only way we can make this work, it is needed in calculations

Reply:
that list defines at what distance we switch from "current" to "historical" endpoint. this is how we do on other exchanges too (i.e. okx ...)

Reply:
If you keep it in options and the exchange decides to update those values it's easier to override on the user side

Reply:
yes, I had moved them in options already.